### PR TITLE
Structural/hcf

### PR DESCRIPTION
**Pull request by SergioJimenezReyes**
Implementing an strategy to study the high cycle fatigue when different cyclic loads are being applied.
I have also changed the code trying to correct the problems mentioned in the issue #5003.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/high_cycle_fatigue_law_integrator.h [hunk 1]
```diff
@@ -89,47 +89,33 @@ class HighCycleFatigueLawIntegrator
...
     {
-        const double stress_1 = rPreviousStresses[1];
-        const double stress_2 = rPreviousStresses[0];
+        const double stress_1 = PreviousStresses[1];
+        const double stress_2 = PreviousStresses[0];
         const double stress_increment_1 = stress_1 - stress_2;
         const double stress_increment_2 = CurrentStress - stress_1;
 
-        if (stress_increment_1 >= 0.001 && stress_increment_2 <= 0.0 && !rCycleCounter) {
+        if (stress_increment_1 >= 0.001 && stress_increment_2 <= 0.0) {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/high_cycle_fatigue_law_integrator.h]: I don't like the hardcoded coefficients 
```suggestion
        if (stress_increment_1 >= 1.0e-3 && stress_increment_2 <= 0.0) {
```

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/high_cycle_fatigue_law_integrator.h [hunk 2]
```diff
@@ -89,47 +89,33 @@ class HighCycleFatigueLawIntegrator
...
             rMaximumStress = stress_1;
-            rPreviousMaximumStress = rMaximumStress;
-            rCycleCounter = true;
-        } else if (stress_increment_1 <= -0.001 && stress_increment_2 >= 0.0 && !rCycleCounter) {
+            rMaxIndicator = true;
+        } else if (stress_increment_1 <= -0.001 && stress_increment_2 >= 0.0) {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/high_cycle_fatigue_law_integrator.h]:
```suggestion
        } else if (stress_increment_1 <= -1.0e-3 && stress_increment_2 >= 0.0) {
```

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp [hunk 3]
```diff
@@ -277,8 +302,8 @@ void GenericSmallStrainHighCycleFatigueLaw<TConstLawIntegratorType>::SetValue(
         GenericSmallStrainIsotropicDamage<TConstLawIntegratorType>::SetValue(rThisVariable, rValue, rCurrentProcessInfo);
     } else if (rThisVariable == FATIGUE_REDUCTION_FACTOR) {
         mFatigueReductionFactor = rValue;
-    } else if (rThisVariable == NUMBER_OF_CYCLES) {
-        mNumberOfCycles = rValue;
+    } else if (rThisVariable == WOHLER_STRESS) {
+        mWohlerStress = rValue;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp:306]:
```suggestion
        noalias(mWohlerStress) = rValue;
```

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp [hunk 4]
```diff
@@ -311,8 +336,8 @@ double& GenericSmallStrainHighCycleFatigueLaw<TConstLawIntegratorType>::GetValue
         GenericSmallStrainIsotropicDamage<TConstLawIntegratorType>::GetValue(rThisVariable, rValue);
     } else if (rThisVariable == FATIGUE_REDUCTION_FACTOR) {
         rValue = mFatigueReductionFactor;
-    } else if (rThisVariable == NUMBER_OF_CYCLES) {
-        rValue = mNumberOfCycles;
+    } else if (rThisVariable == WOHLER_STRESS) {
+        rValue = mWohlerStress;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp:340]:
```suggestion
        noalias(rValue) = mWohlerStress;
```

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.h [hunk 5]
```diff
@@ -321,19 +318,26 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) GenericSmallStrainHighCycleFa
...
     double GetFatigueReductionParameter() {return mFatigueReductionParameter;}
     void SetFatigueReductionParameter(const double toFatigueReductionParameter) {mFatigueReductionParameter = toFatigueReductionParameter;}
 
     Vector GetStressVector() {return mStressVector;}
     void SetStressVector(const Vector toStressVector) {mStressVector = toStressVector;}
 
-    void ResetCycleCounter(){mHasCountedCycle = false;}
-    void SetCycleCounter(const bool tocycle){mHasCountedCycle = tocycle;}
-    bool GetCycleCounter() {return mHasCountedCycle;}
+    bool GetMaxDetected() {return mMaxDetected;}
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.h:334]: Minimal Doxygen please (even if it is obvious)

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minimal comments

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.h (hunk 5 above)
Comment (SergioJimenezReyes) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.h:334]: > Minimal Doxygen please (even if it is obvious)

OK! I'll do it where the variable is declared

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp (hunk 3 above)
Comment (SergioJimenezReyes) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp:306]: mWohlerStress is a double not a vector, 'noalias()' is still required?

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp (hunk 4 above)
Comment (SergioJimenezReyes) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp:340]: same

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp:340]: Nope, my bad, I though

**Review by loumalouomega — APPROVED** [review]
OK

**Review by AlejandroCornejo — APPROVED** [review]
ok!

**Pull request merged**